### DISCUSSION TITLE
Deno  - Getting error "[uncaught application error]: ClientResponseError 0 - Something went wrong while processing your request"

### DISCUSSION THREAD

**Discussion by aym-i** [pocketbase/pocketbase#5365] [Q&A] [answered]
Hi, 
I'm new to pocketbase and wanted to start using it with Deno but whenever I query something I get this error:
```[uncaught application error]: ClientResponseError 0 - Something went wrong while processing your request.```

### I tried:
- `.autoCancellation(false)`
- Checking the Admin UI logs (found nothing)
- `console.log(error)` which outputs: `{"url":"","status":0,"response":{},"isAbort":false,"originalError":{},"name":"ClientResponseError 0"}`
### Code:
```ts
import { Application, Router } from "@oak/oak"
import Pocketbase from "npm:pocketbase"

// This package was installed after I got the error hoping it might be fixed
import { Pbd } from "@ventgrey/pbdq"

// Hardcoding the URL doesn't change anything
const pb = new Pocketbase(Deno.env.get("SERVER_URL"))
const pbd: Pbd = new Pbd({ client: pb })


const router = new Router()
router
    .get("/registry/:ghRepoName", async (context) => {
        // Trying it without filters also fails
        const requestedPackage = await pbd.getFullList({
            collectionName: "packagesRegistry",
            options: { filters: `ghRepoName = ${context.params.ghRepoName}` }
        })            
        context.response.body = requestedPackage
	})

const app = new Application()
app.use(router.routes())
app.use(router.allowedMethods())

await app.listen({ port: 8000 })

```

**Comment by ganigeorgiev** [comment, marked as answer]
`ClientResponseError 0` usually indicates a client-side error, aka. the request has failed to reach the PocketBase server for some reason. This could be caused by incorrect initialization url or simply because the Deno runtime is not allowed/can't access your PocketBase application.

To inspect the original error you could try to log `console.log(error.originalError)`.

**NB!** This will be documented better in the future, but as mentioned in the notes in [JS SDK - LocalAuthStore](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#localauthstore-default) keep in mind that Deno has both `window` and `window.LocalStorage` implemented on the server but its persistence differs from the browser (check the [Deno docs](https://docs.deno.com/runtime/manual/runtime/web_storage_api/) for the specific rules and options). To disable this and to avoid accidentally mixing the regular user auth states you can create your own "blank" store that doesn't persist anything and keeps the auth data only within the created object (this is similar to how it behaves when the JS SDK is used in Node). For example:

```js
import PocketBase, { BaseAuthStore } from "pocketbase"

// extending the base abstract class should be sufficient
class MemoryAuthStore extends BaseAuthStore {}

const pb = new PocketBase("http://127.0.0.1:8090", new MemoryAuthStore())
```

Additionally, if you are planning to use the JS SDK in a JS SSR context, please make sure to check https://github.com/pocketbase/pocketbase/discussions/5313.

**Reply by aym-i** [reply]
Hey, thanks for the response, this is just a simple API, I tried what you told me but I still get the error, here's the full error: 
```
request: {
  url: "http://localhost:8000/registry/test",
  method: "GET",
  hasBody: false
}
response: { status: 500, type: "text", hasBody: true, writable: true }

    at file:///C:/Users/Retcy/AppData/Local/deno/npm/registry.npmjs.org/pocketbase/0.21.4/dist/pocketbase.es.mjs:1:32667
    at eventLoopTick (ext:core/01_core.js:168:7)
    at async file:///C:/Users/Retcy/Nextcloud/Programming/fern/fern-api/src/main.ts:11:33
    at async dispatch (https://deno.land/x/oak@v16.1.0/middleware.ts:90:7)
    at async dispatch (https://deno.land/x/oak@v16.1.0/middleware.ts:90:7)
    at async dispatch (https://deno.land/x/oak@v16.1.0/middleware.ts:90:7)
    at async Application.#handleRequest (https://deno.land/x/oak@v16.1.0/application.ts:605:9)```

EDIT: This may be related to the middleware I'm using, but I don't know

**Reply by ganigeorgiev** [reply]
I'm confused and not sure that I understand what the above error is supposed to show.

>  I tried what you told me ...

Without showing the end code I have no way to confirm this.

Again, if you want to find out the the actual error behind `ClientResponseError 0` you need to log `console.log(error.originalError)` (_note that this is not the same as `console.log(error)` because the parent error may not print/serialize the child one_).

**Reply by aym-i** [reply]
Hey I can confirm that the error output shown is from `error.originalError`

I'll try again and see if it does gives any more details

**Reply by ganigeorgiev** [reply]
I'm not able to reproduce your error. 

I've just installed Deno 1.45.5 and it works correctly for me when I try to access the `/test` endpoint.

There is something else going on in your setup that you didn't show but I don't think it is PocketBase related and I'm not willing to invest more time on this until proven otherwise.

```js
import { Application, Router } from "https://deno.land/x/oak/mod.ts"
import Pocketbase, { BaseAuthStore } from "npm:pocketbase"

// extending the base abstract class should be sufficient
class MemoryAuthStore extends BaseAuthStore {}

// it is safe to be global/shared for all routes in a long running process ONLY when used as admin
// please see "Problem 1" in https://github.com/pocketbase/pocketbase/discussions/5313
const pb = new Pocketbase("https://pocketbase.io", new MemoryAuthStore())

// optional auth as admin
// ---
pb.autoCancellation(false); // we need to be able to process async requests from different users
// the credentials can be filled via env variables
await pb.admins.authWithPassword("test@example.com", "123456", {
  // This will trigger auto refresh or auto reauthentication in case
  // the token has expired or is going to expire in the next 30 minutes.
  autoRefreshThreshold: 30 * 60
})
// ---

const router = new Router()

router.get("/test", async (ctx) => {
    try {
        const posts = await pb.collection("posts").getFullList()
        ctx.response.body = posts
    } catch (err) {
        console.log(err, err.originalError)
        ctx.throw(500);
    }
})

const app = new Application()
app.use(router.routes())
app.use(router.allowedMethods())

await app.listen({ port: 8000 })
```

_edit: updated the minimal example to show instantiating the custom store and authenticating as admin_

**Reply by aym-i** [reply]
Hi!
An update, I no longer get the original error, now I get this one:
I'm I supposed to be authenticated?
```
TypeError: Invalid URL: '/api/collections/packagesRegistry/records?page=1&perPage=500&skipTotal=1'
    at getSerialization (ext:deno_url/00_url.js:98:11)
    at new URL (ext:deno_url/00_url.js:405:27)
    at new Request (ext:deno_fetch/23_request.js:329:25)
    at ext:deno_fetch/26_fetch.js:319:27
    at new Promise (<anonymous>)
    at fetch (ext:deno_fetch/26_fetch.js:315:18)
    at Client.send (file:///C:/Users/me/AppData/Local/deno/npm/registry.npmjs.org/pocketbase/0.21.4/dist/pocketbase.es.mjs:1:32447)
    at RecordService.getList (file:///C:/Users/me/AppData/Local/deno/npm/registry.npmjs.org/pocketbase/0.21.4/dist/pocketbase.es.mjs:1:8239)
    at RecordService.getList (file:///C:/Users/me/AppData/Local/deno/npm/registry.npmjs.org/pocketbase/0.21.4/dist/pocketbase.es.mjs:1:19819)
    at request (file:///C:/Users/me/AppData/Local/deno/npm/registry.npmjs.org/pocketbase/0.21.4/dist/pocketbase.es.mjs:1:9529)
```
Code:
```ts
import { Application, Router } from "https://deno.land/x/oak/mod.ts"
import Pocketbase, { BaseAuthStore } from "npm:pocketbase"

class MemoryAuthStore extends BaseAuthStore {}

const pb = new Pocketbase("https://fern-a.pockethost.io", new MemoryAuthStore())

const router = new Router()
router
    .get("/registry/:ghRepoName", async (context) => {
        console.log("HI", Deno.env.get("SERVER_URL"))
        context.response.body = await pb.collection("packagesRegistry").getFullList()
     })

const app = new Application()
app.use(router.routes())
app.use(router.allowedMethods())

await app.listen({ port: 8000 })
```
API Rules of the table:
![image](https://github.com/user-attachments/assets/c93c5a43-48a6-4c94-a146-20011a877f34)

**Reply by ganigeorgiev** [reply]
@retr0cube I've tested your example and it works correctly for me with deno 1.45.5 (the only thing I've changed was the PocketBase server url).

Again, I don't think your isssue is PocketBase related and there is probably something else going on in your environment that you didn't share but without clear reproducible steps I can't help you further.

**Reply by aym-i** [reply]
> @retr0cube I've tested your example and it works correctly for me with deno 1.45.5 (the only thing I've changed was the PocketBase server url).
> 
> Again, I don't think your isssue is PocketBase related and there is probably something else going on in your environment that you didn't share but without clear reproducible steps I can't help you further.

Can you use the same URL as mine it only has one entry and public anyways?

**Reply by ganigeorgiev** [reply]
Done. I've tested the example exactly as you've pasted it and run it with `deno run index.js` (I've granted access to `0.0.0.0:8000`, your url and the env variable read). When accessing `http://localhost:8000/registry/test` I got:

```js
[{"collectionId":"plwkbqpue6dq2ux","collectionName":"packagesRegistry","created":"2024-08-16 18:03:43.412Z","description":"Add-on development kit for Fern, currently in development.","downloads":0,"ghRepoAuthor":"retr0cube","ghRepoName":"fern-adk","id":"75y217hakaq65mt","lastUpdateDate":"","links":null,"name":"Fern Add-on Development Kit","publishDate":"2024-08-16 19:00:00.000Z","tags":"official,tool","type":"tool","updated":"2024-08-16 18:03:43.412Z"}]
```

Note that if you are planning to use the PocketBase JS SDK as admin I recommend the example from https://github.com/pocketbase/pocketbase/discussions/5365#discussioncomment-10363018.

I'll stop engaging as I think I wasted enough time here. Please understand that this is not a general dev support forum and you may find better help for example in StackOverflow or in some of the Deno support channels.

**Reply by aym-i** [reply]
> Done. I've tested the example exactly as you've pasted it and run it with `deno run index.js` (I've granted access to `0.0.0.0:8000`, your url and the env variable read). When accessing `http://localhost:8000/registry/test` I got:
> 
> ```js
> [{"collectionId":"plwkbqpue6dq2ux","collectionName":"packagesRegistry","created":"2024-08-16 18:03:43.412Z","description":"Add-on development kit for Fern, currently in development.","downloads":0,"ghRepoAuthor":"retr0cube","ghRepoName":"fern-adk","id":"75y217hakaq65mt","lastUpdateDate":"","links":null,"name":"Fern Add-on Development Kit","publishDate":"2024-08-16 19:00:00.000Z","tags":"official,tool","type":"tool","updated":"2024-08-16 18:03:43.412Z"}]
> ```
> 
> Note that if you are planning to use the PocketBase JS SDK as admin I recommend the example from [#5365 (reply in thread)](https://github.com/pocketbase/pocketbase/discussions/5365#discussioncomment-10363018).
> 
> I'll stop engaging as I think I wasted enough time here. Please understand that this is not a general dev support forum and you may find better help for example in StackOverflow or in some of the Deno support channels.

Hey thanks for the support, I really appreciate it, will investigate further and find the solution, but the error is from PB? I don't understand

**_Update: It works now, just needed to restart the server, so weird... it turns out the env variable was returning undefined :/_**
